service/history/workflow/mutable_state_impl.go

```diff
@@ -938,30 +943,63 @@ func (ms *MutableStateImpl) GetStartVersion() (int64, error) {
 		if err != nil {
 			return 0, err
 		}
-		firstItem, err := versionhistory.GetFirstVersionHistoryItem(versionHistory)
-		if err != nil {
-			return 0, err
+
+		if !versionhistory.IsEmptyVersionHistory(versionHistory) {
+			firstItem, err := versionhistory.GetFirstVersionHistoryItem(versionHistory)
+			if err != nil {
+				return 0, err
+			}
+			return firstItem.GetVersion(), nil
 		}
-		return firstItem.GetVersion(), nil
+	}
+
+	// We can't check TransitionHistory before VersionHistories because old workflows (mutable state) may
+	// have transition history enabled while they are running, so the first item in the transition history
+	// is not the actual start version.
+	if ms.transitionHistoryEnabled && len(ms.executionInfo.TransitionHistory) != 0 {
+		return ms.executionInfo.TransitionHistory[0].GetNamespaceFailoverVersion(), nil
 	}
 
 	return common.EmptyVersion, nil
 }
 
 func (ms *MutableStateImpl) GetCloseVersion() (int64, error) {
-	if ms.IsWorkflowExecutionRunning() {
+	// Do NOT use ms.IsWorkflowExecutionRunning() for the check.
+	// Zombie workflow is not considered running but also not closed.
+	if ms.executionState.State != enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED {
```

Comment:
I need to double check all the callers. This is breaking ndc workflow right now, which may call it when workflow is in zombie state.

Reply:
Guaranteeing caller not calling this when workflow is zombie will require a big refactoring. For now, I special handled zombie case and returns LastWriteVersion instead.